### ISSUE TITLE
Particle creation is slow in parallel simulations

### ISSUE THREAD

**Issue by RudolfWeeber** [espressomd/espresso#5011]
The time for creating a particle scales linearly in the number of mpi ranks for high core numbers (>~64)
E.g 30 ms per particle creatoin (explicitly specifying the particle id) for 512 MPi ranks on the ant cluster at ICP.

In my opionion, particle insertoin should comprise
* a broadcast from Python on rank 0 to all ranks (scales n log n in number of ranks)
* maybe some point to point communicaoitn at constatn time
## Steps to investigate

Use a Python profiler on the particle creation code in src/python/espressomd.particle_data.py with different number of mpi ranks to figure out, where in the creation process the poor scaling appears.

The script was

```python
import espressomd
import numpy as np
from time import time

s=espressomd.System(box_l=[1,1,1])
grid =s.cell_system.node_grid
n=10*np.prod(grid)
print(grid)
for n in 1,10,100:
    tick=time()
    s.part.add(id=range(n),pos=np.random.random((n,3))*s.box_l)
    tock=time()
    print(n, (tock-tick)/n)
    s.part.clear()
```

And the commands to run were

```bash
module load spack/default gcc/12.3.0 cuda/12.3.0 openmpi/4.1.6 \
            fftw/3.3.10 boost/1.83.0 cmake/3.27.9 python/3.12.1

`
srun --cpu-bind=cores --ntasks=512 --time=60 -u ./pypresso ./test.py 
```

**Comment by jngrad** [comment]
I also noticed excessive particle insertion times on the HPC Vega supercomputer. This issue can be investigated using a python debugger and/or profiler. See user documentation section on  [instrumentation](https://espressomd.github.io/doc/running.html#instrumentation) for natively supported tools.

**Comment by RiccardoFrenner** [comment]
# First results
Okay, it is clear to me why **Rudolf's** script has O(nodes) scaling. The function [`mpi_who_has_head`](https://github.com/espressomd/espresso/blob/91fa77acc6794f6f0530824f7b31dbd4b87daf9f/src/core/particle_node.cpp#L305) which is called once at the beginning (on rank 0) and also after the particles are cleared, iterates through all nodes and performs communication with each node to setup the p_id -> node_id mapping. So the more nodes there are, the slower `mpi_who_has_head` is.

However, I still get a linear scaling without `mpi_who_has_head` being called as demonstrated by the below script and the plots:

```python
from time import time

import sys
import espressomd
import numpy as np

s = espressomd.System(box_l=[1, 1, 1])
grid = s.cell_system.node_grid
n_cores = np.prod(grid)

# without this, the first particle insertion would be significantly slower, unrelated to n_cores or n_particles
s.part.add(id=list(range(1000)), pos=np.random.random((1000, 3)) * s.box_l)
s.part.clear()

print("-"*100)

for n in [1, 10, 100]:
    ids = np.arange(n)
    # this causes mpi_who_has_head to be called
    dt_first = time()
    s.part.add(id=ids, pos=np.random.random((n, 3)) * s.box_l)
    dt_first = time() - dt_first
    
    print("-"*100)

    # here mpi_who_has_head is not called again
    dt_second = time()
    s.part.add(id=ids + n, pos=np.random.random((n, 3)) * s.box_l)
    dt_second = time() - dt_second

    s.part.clear()

    print(n_cores, n, dt_first, dt_second, sep=",", file=sys.stderr)
```

<img width="1022" alt="Image" src="https://github.com/user-attachments/assets/f14ec3be-2492-44ad-b22c-52ceab4b2575" />

I am aware that you were talking about much more than 10 cores which was the most I used in this experiment (resulting in a latency of 0.5 ms instead of your 30 ms), but I was still seeing a linear scaling here which should not be there (correct me if I am wrong). Soon I will have an ant account and can redo this with more cores there.


# Questions
The above is just what I determined from looking at the code and running experiments.

I did not manage to use kernprof with more than one MPI rank. Any idea how to do this?
Running on a single rank just got me that `part.by_id` is the bottleneck of `part.add`. But `by_id` calls into the c++ code directly, and kernprof does not show me the timings for our c++ functions.

If I now wanted to get the timings of the c++ functions but with more than one mpi rank, how would I do this? Is `perf` the right tool or rather `Callgrind`, because it says `perf` requires root?

Also, it would be good to know what exactly the use case is. Because if it is about adding particles, then doing some computations (integrating, computing observables, etc.) and repeating this, this will always cause `mpi_who_has_head` to be called after the computation since the p_id -> node_id map will be cleared, causing a rebuild. So in this case, adding particles will always scale with the number of nodes.
If nothing like this is done in between adding particles (not sure about the use case here; this is the case that I measure with `dt_second` in my example), further investigation is necessary.

**Comment by RudolfWeeber** [comment]
Thanks, Riccardo. Ant account should be ready now. Concerning your questoins:
* I don't think, kernprof has been used so far. I'd try loading it from within particle_data.py rather than  from the command line using mpirun
* So, under the conditoin that particle ids are specified, when adding particles, can the poor scaling be linked to a specific call to the core (as there is quite a bit of Python code for particle creation)?
* Timings in C++ are best done using CALIPER. For that, one basically hsa to mark the functions for which one needs timings. You'll find exapmles in the core, when grepping for CALI. I'm not sure, if the CMake configuraiotn for Caliper (it has to be turned on, there), is visible in the script interface (@jngrad)

**Comment by jngrad** [comment]
> You'll find exapmles in the core, when grepping for CALI. I'm not sure, if the CMake configuraiotn for Caliper (it has to be turned on, there), is visible in the script interface ([@jngrad](https://github.com/jngrad))

It's exposed to the script interface.

**Comment by RiccardoFrenner** [comment]
# Python Profiling

> So, under the conditoin that particle ids are specified, when adding particles, can the poor scaling be linked to a specific call to the core (as there is quite a bit of Python code for particle creation)?

Yes, the time spent is almost equally divided between `self.call_method("add_particle", **p_dict)` and `self.call_method("by_id", p_id=p_id)`, although for increasing core count, the ratio shifts towards `add_particle`. All other methods have minimal impact and no scaling issues. Below is an example output for 64 cores using the script at the bottom of this comment:
<details>
  <summary>Python Profiler Output</summary>
  

```
Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
  1123                                               def _place_new_particle(self, p_dict):
  1124      1222     667435.0    546.2      0.0          bonds = []
  1125      1222     385164.0    315.2      0.0          if "propagation" in p_dict:
  1126                                                       p_dict["propagation"] = int(p_dict["propagation"])
  1127      1222     247872.0    202.8      0.0          if "bonds" in p_dict:
  1128                                                       bonds = p_dict.pop("bonds")
  1129                                                       if nesting_level(bonds) == 1:
  1130                                                           bonds = [bonds]
  1131      1222 1122807966.0 918828.1     75.8          p_id = self.call_method("add_particle", **p_dict)
  1132      1222  355863289.0 291213.8     24.0          p = self.by_id(p_id)
  1133      1222     859542.0    703.4      0.1          for bond in bonds:
  1134                                                       if len(bond):
  1135                                                           bond = p.normalize_and_check_bond_or_throw_exception(bond)
  1136                                                           p.add_verified_bond(bond)
  1137      1222     472597.0    386.7      0.0          return p
```
</details>
(When profiling `s.part.add`, 99% of time was spent in `_place_new_particle`.)

I repeated the same for various core counts and could see a linear increase with increasing core count in time both in `add_particle` and in `by_id`.

So we can now say that the python code is not the culprit because both functions directly call into the c++ code via `self.call_method`.


# CALIPER

From using CALIPER, I suspect `do_set_parameter` to be the culprit. I then tried to examine the `AutoParameter::set` method, specifically for the `"pos"` parameter, because this one seems more likely than the `"id"` parameter. However, since `AutoParameter` takes lambdas, putting the CALI macros inside the actual function that the lambda calls, `set_particle_pos`, does not work. I also tried putting the macro inside the lambda, but this did not work either. So I cannot determine if it really is the "pos" setter that is the problem and in that case what exactly inside it is the problem.

Any ideas how to proceed?



<details>
  <summary>Script used for Python Profiling with MPI</summary>
  
```python
from time import time
import sys
import espressomd
import numpy as np
from line_profiler import LineProfiler

def run_profile():
    s = espressomd.System(box_l=[1, 1, 1])
    grid = s.cell_system.node_grid
    n_cores = np.prod(grid)

    s.part.add(id=list(range(1000)), pos=np.random.random((1000, 3)) * s.box_l)
    s.part.clear()

    for n in [1, 10, 100]:
        ids = np.arange(n)
        dt_first = time()
        s.part.add(id=ids, pos=np.random.random((n, 3)) * s.box_l)
        dt_first = time() - dt_first

        dt_second = time()
        s.part.add(id=ids + n, pos=np.random.random((n, 3)) * s.box_l)
        dt_second = time() - dt_second

        s.part.clear()

        print(n_cores, n, dt_first, dt_second, sep=",")#, file=sys.stderr)

if __name__ == '__main__':
    profiler = LineProfiler()
    # profiler.add_function(espressomd.particle_data.ParticleList.add)
    # profiler.add_function(espressomd.particle_data.ParticleList._place_new_particles)
    profiler.add_function(espressomd.particle_data.ParticleList._place_new_particle)
    profiler.runcall(run_profile)
    profiler.print_stats()
```
</details>

**Comment by RiccardoFrenner** [comment]
I also looked at the timings of MPI functions using CALIPER and found MPI_Isend, MPI_Test, MPI_Improbe (together take 4% of total time) scale the worst (O(1.5 n)) with increasing core count. However, MPI_Mprobe takes the most amount of total time (70%) and scales O(0.5 n), MPI_Allreduce takes 16% of the time and also scales O(0.5 n).

Not sure what to make from this though.

**Comment by RudolfWeeber** [comment]
Thanks for the update, Riccardo. 
Here are some things you might explore (in order of my prediction of effect per effort):
1. in Python in _place_new_particle, only construct the ParticleHandle, if bonds are actually to be added. Do not return the newly constucted ParticleHandle from _place_new_particle(). Instead, for a single particle, create the instance in system.part.add(). Similarly for adding many particles: return the list of ids form _place_new_particles and instance a ParticleSlice object in ParticleList.add()
2. Add a skip_validation flag to the parameters of ParticleHandle::do_constfuct in the script interface. If true, the constructor can assume that the particle exists and the id is valid and skip most of the parallel calls in the constructor. Make use of this flag in ParticleList._place_new_particle, and ParticleList.add().

**Comment by RiccardoFrenner** [comment]
# 1. Simplifying the python method system.part.add
I am not sure if I understood this correctly, but I now changed `_place_new_particle` to
```python
def _place_new_particle(self, p_dict):
        p_id = self.call_method("add_particle", **p_dict)
```
and call it directly via `s.part._place_new_particle(dict(id=i, pos=pos[i]))` in my test script instead of `s.part.add`. The line profiling then also reduces to a single line:
```
Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
  1137                                               def _place_new_particle(self, p_dict):
  1138      1200   87313203.0  72761.0    100.0          p_id = self.call_method("add_particle", **p_dict)
```

But I still get a linear time increase with increasing core count for 64 cores and more (also for more than 1 particle added):
```
n_cores duration(s)
256	0.007822
128	0.005040
64	0.002570
32	0.000160
16	0.000117
8	0.000091
4	0.000081
2	0.000071
1	0.000050
```

This again shows that the issue is the c++ code, not the python code. Please let me know if I did not implement your test idea correctly.


# 2. Simplifying the c++ methods

I commented out a few parts of the constructor, specifically, the `parallel_try_catch` where `particle_checks` is called and the `parallel_try_catch` inside the ROTATION region. I don't think there are more checks that could be skipped. Is this what you had in mind?

I did not comment out
- `get_maximal_particle_id` since is not called because `params` contains "id".
- and the last `parallel_try_catch` cannot be skipped because it sets all parameters except `pos` and does no other checks directly here.

But I still got the same timings and scaling issues (actually slightly improved for <= 128 cores but not significantly).

<details>
  <summary>Modified ParticleHandle::do_construct method</summary>

```c++
void ParticleHandle::do_construct(VariantMap const &params) {
  auto const n_extra_args = params.size() - params.count("id") -
                            params.count("__cell_structure") -
                            params.count("__bonded_ias");
  m_pid = (params.contains("id")) ? get_value<int>(params, "id")
                                  : get_maximal_particle_id() + 1;

#ifndef NDEBUG
  if (not params.contains("id")) {
    auto head_node_reference = m_pid;
    boost::mpi::broadcast(context()->get_comm(), head_node_reference, 0);
    assert(m_pid == head_node_reference && "global max_seen_pid has diverged");
  }
#endif

  if (params.contains("__cell_structure")) {
    auto so = get_value<std::shared_ptr<CellSystem::CellSystem>>(
        params, "__cell_structure");
    so->configure(*this);
    m_cell_structure = so;
  }
  if (params.contains("__bonded_ias")) {
    m_bonded_ias = get_value<std::shared_ptr<Interactions::BondedInteractions>>(
        params, "__bonded_ias");
  }

  // create a new particle if extra arguments were passed
  if (n_extra_args == 0) {
    return;
  }

  auto const pos = get_value<Utils::Vector3d>(params, "pos");
  // context()->parallel_try_catch([&]() {
  //   particle_checks(m_pid, pos);
  //   auto cell_structure_si = get_cell_structure();
  //   auto &cell_structure = cell_structure_si->get_cell_structure();
  //   auto ptr = cell_structure.get_local_particle(m_pid);
  //   if (ptr != nullptr) {
  //     throw std::invalid_argument("Particle " + std::to_string(m_pid) +
  //                                 " already exists");
  //   }
  // });

// #ifdef ROTATION
//   context()->parallel_try_catch([&]() {
//     // if we are not constructing a particle from a checkpoint file,
//     // check the quaternion is not accidentally set twice by the user
//     if (not params.contains("__cpt_sentinel")) {
//       auto formatter =
//           boost::format("Contradicting particle attributes: '%s' and '%s'. %s");
//       for (auto const &[prop1, prop2, reason] : contradicting_arguments_quat) {
//         if (params.contains(prop1) and params.contains(prop2)) {
//           auto const err_msg = boost::str(formatter % prop1 % prop2 % reason);
//           throw std::invalid_argument(err_msg);
//         }
//       }
//     }
//   });
// #endif // ROTATION

  // create a default-constructed particle
  make_new_particle(m_pid, pos);

  try {
    context()->parallel_try_catch([&]() {
      /* clang-format off */
      // set particle properties (filter out read-only and deferred properties)
      std::set<std::string> const skip = {
          "pos_folded", "pos", "id", "exclusions", "node", "image_box", "bonds",
          "lees_edwards_flag", "__cpt_sentinel",
      };
      /* clang-format on */
      for (auto const &name : get_parameter_insertion_order()) {
        if (params.contains(name) and not skip.contains(name)) {
          do_set_parameter(name, params.at(name));
        }
      }
      if (not params.contains("type")) {
        do_set_parameter("type", 0);
      }
    });
  } catch (...) {
    remove_particle(m_pid);
    throw;
  }
}
```

</details>

**Comment by RudolfWeeber** [comment]
OK, not sure I got everything correclty. Maybe we should discss in person.
Some points:
* If I understand correctly, we would get a 50% speedup (but still a linear scaling) if we avoided the by_id method call. As I understand it, the script interface method ParticleList::add_particle anyway constrcuts the partilce object, but only returns the id to python. Can we just return the script interface ParticleHanlde object instanced in the add_particle method to Python. Would this let us avoid the consecutive get particle by id call?
* It would be helpful to understand the scaling of script interface "primitives" such as a) a method call on an existing script interface object, b) the instancing of a new script interface object, c) a parallel try-catch. Then, we could address the actual scalability problem.

**Comment by RudolfWeeber** [comment]
Ah, your modified _place_new_particle already avoids the by_id call. So, is that version actually faster?
For creating many particles, is the get_by_ids call at the very end of _place_new_particles instancing the particle slice actually an issue?

**Comment by RiccardoFrenner** [comment]
Yes, let's meet in person.


> Ah, your modified _place_new_particle already avoids the by_id call. So, is that version actually faster?

Only for small core counts. Details in the summary at the end.

> For creating many particles, is the get_by_ids call at the very end of _place_new_particles instancing the particle slice actually an issue?

Maybe it can be optimized since `_place_new_particle` already returns an ID. But I don't call `_place_new_particles` in my tests (I just loop `_place_new_particle`), so it is unrelated for this scaling issue here.

> ParticleList::add_particle anyway constrcuts the partilce object, but only returns the id to python. Can we just return the script interface ParticleHanlde object instanced in the add_particle method to Python. Would this let us avoid the consecutive get particle by id call?

Not really sure if I understood this correctly. The ParticleHandle object is constructed but not returned. Instead the ID is returned, that is correct. But the ID is just obtained by calling the `get_parameter` member function of the ParticleHandle object. I don't think this function scales badly. I think the ID is already set (which may be expensive) at construction.
If you are talking about the python code, so the `add` function of the system class, yes there is the add_particle function after which the by_id function is called. So this could maybe be optimized. But in my test code, this is not used but the simplified `_place_new_particle` function. So this optimization does not fix the issue.


# Summary of so-far attempted Optimizations

- Before any changes: 8ms latency for 256 cores and 1 particle
- After modifying python `_place_new_particle` and removing c++ checks: 7ms latency for 256 cores and 1 particle
- So a delta of 1ms: `delta@256c = 1ms`

As mentioned above, the latency between `by_id` and `add_particle` is only 50/50 for small core counts. For higher this shifts: e.g. 64c -> 25%/75%.
This is in * with the delta at smaller core counts being larger, e.g. `delta@64c = 1.5ms`.
